### PR TITLE
Change `F.dropout` to use cuDNN by default

### LINKED ISSUE #6266: Why `F.dropout` usually does not use CuDNN?
Currently, `F.dropout` uses cudnn only when `using_config('use_cudnn', 'always')`
https://github.com/chainer/chainer/blob/d7adbfb45d46fafa6bd0bbb9f0cebe5894489437/chainer/functions/noise/dropout.py#L46

But IMO, cudnn should be enabled when `>=auto`.

### PR DESCRIPTION
**Pull request by crcrpar**
Close #6266

### CODE REVIEW

File: tests/chainer_tests/functions_tests/noise_tests/test_dropout.py [hunk 1]
```diff
@@ -183,7 +184,8 @@ def test_cpu_ideep(self):
     def test_gpu(self):
         self.x = cuda.to_gpu(self.x)
         self.mask = cuda.to_gpu(self.mask)
-        self._check()
+        with chainer.using_config('use_cudnn', 'never'):
+            self._check()
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: How about adding a new test for `use_cudnn='always'` (as `test_gpu_cudnn`)?
Perhaps it's better to use `inject_backend_tests`.

Comment (crcrpar) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: I added tests, but it is possibly implemented in a wrong way.

File: tests/chainer_tests/functions_tests/noise_tests/test_dropout.py [hunk 2]
```diff
@@ -144,46 +148,49 @@ def test_immutable(self, backend_config):
...
-        mask = self.mask if self.specify_mask else None
-        with chainer.using_config('train', self.train):
-            out, out_mask = functions.dropout(
-                self.x, 0.5, mask=mask, return_mask=True)
+    def test_check(self, backend_config):
+        mask = None
+        if self.specify_mask:
+            mask = self.mask
 
+        x, mask = backend_config.get_array((self.x.copy(), mask))
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: No need to `copy()` I guess.

File: tests/chainer_tests/functions_tests/noise_tests/test_dropout.py (hunk 1 above)
Comment (niboshi) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: Thanks!

File: tests/chainer_tests/functions_tests/noise_tests/test_dropout.py [hunk 3]
```diff
@@ -144,46 +148,49 @@ def test_immutable(self, backend_config):
...
+        with chainer.using_config('train', self.train), backend_config:
+            out, out_mask = functions.dropout(
+                x, 0.5, mask=mask, return_mask=True)
         if self.train:
-            assert isinstance(out_mask, type(out.array))
-            if mask is None:
+            if mask is None and backend_config.use_cudnn != 'always':
+                assert isinstance(out_mask, type(out.array))
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: Shouldn't this assertion always satisfied?

Comment (niboshi) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: Sorry, I mean it should be always checked.

Comment (crcrpar) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: How can we get mask from cudnn implementation?

Comment (crcrpar) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]:
```bash
Platform: Linux-4.9.0-8-amd64-x86_64-with-debian-9.8
Chainer: 7.0.0a1
NumPy: 1.16.2
CuPy:
  CuPy Version          : 7.0.0a1
  CUDA Root             : /usr/local/cuda
  CUDA Build Version    : 10000
  CUDA Driver Version   : 10000
  CUDA Runtime Version  : 10000
  cuDNN Build Version   : 7401
  cuDNN Version         : 7401
  NCCL Build Version    : None
  NCCL Runtime Version  : None
iDeep: Not Available

In [2]: x = cp.random.uniform(-1, 1, (10, 32)).astype(cp.float32)                                                                                                  
In [3]: with chainer.using_config('use_cudnn', 'always'): y, out_mask = F.dropout(x, return_mask=True)                                                             
In [4]: out_mask is None                                                                                                                                           
Out[4]: True
```

Comment (crcrpar) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: How will I handle cudnn enabled forward?

Comment (niboshi) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: Hmm, how about enabling cuDNN only when `not return_mask`?
You will need to pass `return_mask` to the `Dropout` class.

Comment (niboshi) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: I think this is a preexisting bug.
Could you separate a PR, as we need to backport the fix?

Comment (niboshi) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: Or it's ok to just dodge the bug in this PR and fix it at some other time.

Comment (crcrpar) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: OK, I'll send a PR to make `Dropout` FunctionNode consider `return_mask`, then I'll add tests that check cuDNN correctly used/unused.

Is my understanding correct?

Comment (niboshi) [reply tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: Yes. Thank you very much!

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM

**Pull request merged**